node/src/service/nonce_monitor.rs

```diff
@@ -0,0 +1,193 @@
+use substrate_prometheus_endpoint::Registry;
+use thiserror::Error;
+
+use super::FullBackend;
+
+#[derive(Clone, Copy, PartialEq, Debug, Error)]
+enum ErrorKind {
+	#[error("Serde error")]
+	Serde,
+	#[error("RPC error")]
+	Rpc,
+	#[error("Codec error")]
+	Codec,
+}
+
+#[derive(Debug, Error)]
+#[error("{kind}: {value}")]
+struct Error {
+	kind: ErrorKind,
+	#[source]
+	value: Box<dyn std::error::Error + Send>,
+}
+
+macro_rules! error_fn {
+	($($fn_id: ident : $kind: ident),+ $(,)?) => {
+		impl Error {
+			$(
+				fn $fn_id(err: impl std::error::Error + Send + 'static) -> Self {
+					Self::new(ErrorKind::$kind, err)
+				}
+			)+
+		}
+	};
+}
+
+error_fn!(serde: Serde, rpc: Rpc, codec: Codec);
+
+impl Error {
+	fn new(kind: ErrorKind, err: impl std::error::Error + Send + 'static) -> Self {
+		Self { kind, value: Box::new(err) }
+	}
+}
+
+#[derive(Clone, PartialEq, Debug, Error)]
+#[error("{message}")]
+struct AdhocError {
+	message: String,
+}
+
+fn adhoc(message: impl Into<String>) -> AdhocError {
+	AdhocError { message: message.into() }
+}
+
+async fn rpc_request(
+	handlers: &RpcHandlers,
+	request: &str,
+) -> Result<jsonrpc_core::serde_json::Value, Error> {
+	let (tx, _rx) = mpsc::unbounded();
+	let session = RpcSession::new(tx);
+
+	let response = handlers
+		.rpc_query(&session, request)
+		.await
+		.ok_or_else(|| Error::rpc(adhoc("empty response")))?;
+
+	let response: Response = jsonrpc_core::serde_json::from_str(&response).map_err(Error::serde)?;
+
+	let result = match response {
+		Response::Single(out) => match out {
+			jsonrpc_core::Output::Success(Success { result, .. }) => result,
+			jsonrpc_core::Output::Failure(Failure { error, .. }) => return Err(Error::rpc(error)),
+		},
+		Response::Batch(_) => {
+			unreachable!("we don't send any batch requests, so we cannot receive batch responses")
+		},
+	};
+
+	Ok(result)
+}
+
+async fn get_on_chain_nonce(handlers: &RpcHandlers, acct: &AccountId) -> Result<u64, Error> {
+	let request = format!(
+		r#"{{
+            "jsonrpc": "2.0",
+            "method": "system_accountNextIndex",
+            "params": ["{}"],
+            "id": 0
+        }}"#,
+		acct.to_ss58check()
+	);
+
+	let result = rpc_request(handlers, &request).await?;
+
+	result.as_u64().ok_or_else(|| Error::rpc(adhoc("expected u64 response")))
+}
+
+async fn get_off_chain_nonce_key(
+	handlers: &RpcHandlers,
+	acct: &AccountId,
+) -> Result<Vec<u8>, Error> {
+	let request = format!(
+		r#"{{
+            "jsonrpc": "2.0",
+            "method": "task_getOffchainNonceKey",
+            "params": ["{}"],
+            "id": 0
+        }}"#,
+		acct.to_ss58check()
+	);
+
+	let result = rpc_request(handlers, &request).await?;
+
+	let key: Vec<u8> = jsonrpc_core::serde_json::from_value(result).map_err(Error::serde)?;
+
+	Ok(key)
+}
+
+async fn get_off_chain_nonce(backend: &FullBackend, key: &[u8]) -> Result<Option<u64>, Error> {
+	let off = backend.offchain_storage().unwrap().get(sp_offchain::STORAGE_PREFIX, key);
```

Comment:
is this `unwrap` safe?

Reply:
Hmm the only way it could fail is if offchain storage is completely missing, which I believe is only the case for testing environments